csharp/src/Ice/Communicator.cs

```diff
@@ -1364,5 +1273,106 @@ private void AddAllAdminFacets()
 
             return null;
         }
+
+        private async Task PerformDisposeAsync()
+        {
+            lock (_mutex)
+            {
+                Debug.Assert(_state < StateDestroyInProgress);
+                _state = StateDestroyInProgress;
+            }
+
+            // Cancel operations that are waiting and using the communicator's cancellation token
+            _cancellationTokenSource.Cancel();
+
+            // Shutdown and destroy all the incoming and outgoing Ice connections and wait for the connections to be
+            // finished.
+            _ = ShutdownAsync();
+
+            _outgoingConnectionFactory?.DestroyAsync();
+
+            await WaitForShutdownAsync().ConfigureAwait(false);
+            _shutdownTaskSemaphore?.Dispose();
```

Comment:
Interesting... I didn't know that semaphore slim need to be disposed! We'll need to fix few usage of semaphore slim to dispose them. That said, according to some threads I read on SO, disposing the semaphore doesn't appear to be useful unless you use the AwaitatableWaitHandle form the semaphore... but probably best to still dispose it anyway.

So I wonder if calling Dispose here is really correct since the documentation states that it's not thread safe and here we can't be sure there other threads waiting on WaitAsync calls on the semaphore slim are all completed... I'm thinking it's probably best to move this to the WaitForShutdownAsync implementation and do something like the following:

```
            try
            {
                await shutdownTask.ConfigureAwait(false);
            }
            finally
            {
                // The last awaiter is responsible for calling Dispose on the semaphore
                if (_shutdownTaskSemaphore?.CurrentCount == 0)
                {
                    _shutdownTaskSemaphore.Dispose();
                }
            }
```